### PR TITLE
[For v0.18]Colors management (further changes ongoing, do not take)

### PR DESCRIPTION
**Pull request by plgarcia**
Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [x] Branch rebased on latest master `git pull --rebase upstream master`
- [x] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.

---

### CODE REVIEW

**Comment by plgarcia** [comment]
https://forum.freecadweb.org/viewtopic.php?f=8&t=25505&sid=ad32163b408874f794f3f962043ab930

related to issue #848

**Comment by plgarcia** [comment]
This PR is the first step for solving the problems of colors of the composed parts.
Il applies for the cut, fusion,  Booleans, compound, champfer and fillet parts. It will be generalize to other composed parts (pockets, etc).
The first commit adds two targets for creating more detailed documentation for py and cpp. It is very minor but helpful to understand better how the software is structured. There are for developers and are space consuming (12 Gb for the cpp doc). These two targets can be used as follow: 

- make -j2 DevDocPy
- make -j2 DevDocCpp

The second commit adds the generic class for derived parts.
The third uses these class for the different parts listed above.
The problems that are not yet solved are:
When changing the color of a part the color is not reported to the parts directly or indirectly composed with this object.
Changing the color to black of a composed part does not work for the objects present in the file before loading the file. This is due to history not saved, and therefore not loaded. This problem disappears when the object is recomputed. When the problem is detected by the code the object is marked to be recomputed, so F5 clears this problem. This can also be solved by marking to recompute the full document and F5, so all structures are recomputed and present in memory.
Coming the generalization to more composed parts.

File: src/Mod/Part/App/AppPart.cpp [hunk 1]
```diff
@@ -424,6 +425,7 @@ PyMOD_INIT_FUNC(Part)
     Part::Primitive             ::init();
     Part::Box                   ::init();
     Part::Spline                ::init();
+	Part::FeatureDerivedPart    ::init();
```
Comment (luzpaz) [inline src/Mod/Part/App/AppPart.cpp]: Any reason for the whitespace change ?

File: src/Mod/Part/App/FeatureDerivedPart.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#define PART_FEATUREDERIVEDPART_H
+
+
+#include "PartFeature.h"
+#include <App/PropertyLinks.h>
+
+namespace Part
+{
+
+// "derived part" is an abstract part witch derive any part such as boolean or fusions result of operations between one of many children objects
```
Comment (luzpaz) [inline src/Mod/Part/App/FeatureDerivedPart.h]: s/witch/which/
also maybe put the comment on 2 lines for better readability?

File: src/Mod/Part/Gui/ViewProviderDerivedPart.cpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+		return;
+	const std::vector<Part::ShapeHistory>& hist = objPart->History.getValues();
+	const TopoDS_Shape& partShape = objPart->Shape.getValue();
+	TopTools_IndexedMapOfShape partMap;
+	TopExp::MapShapes(partShape, TopAbs_FACE, partMap);
+
+	std::vector<App::Color> colPart;
+	
+	if (hist.size() != sources.size()) {
+		// there is no history that should be calcultated while executing the feaure
```
Comment (luzpaz) [inline src/Mod/Part/Gui/ViewProviderDerivedPart.cpp]: s/calcultated/calculated/

**Comment by plgarcia** [comment]
Usage:
When setting one of the composed parts to black (#000000), the colors used for this part are the colors of the composing parts. If the transparency is also set to 0 then the transparencies used are the transparencies of the composing parts instead of the transparency of the part itself.

If a color is set to something different to black, then this colors and the transparency are used for all the part.

File: src/Mod/Part/Gui/ViewProviderDerivedPart.cpp (hunk 3 above)
Comment (plgarcia) [reply src/Mod/Part/Gui/ViewProviderDerivedPart.cpp]: Sorry I will correct that

File: src/Mod/Part/App/AppPart.cpp (hunk 1 above)
Comment (plgarcia) [reply src/Mod/Part/App/AppPart.cpp]: On my editor there is no difference. may be due to the size of the tabs. I will correct that.

File: src/Mod/Part/App/FeatureDerivedPart.h (hunk 2 above)
Comment (plgarcia) [reply src/Mod/Part/App/FeatureDerivedPart.h]: OK I will do it

**Comment by luzpaz** [comment]
PR name should be Color manag**em**ent

**Comment by plgarcia** [comment]
Just to know what to do.
Shall I continue to generalize this to other parts and solve the problems arround colors everywhere?
Shall I change something befor going further as may be the name of the class ...DerivedPart?

Regards
Pascal Garcia

**Comment by plgarcia** [comment]
During development, some problems raised that need transformation of the core.
It happens that in many situation calculations are done twice when a change occurs. 
To evaluate colors the history is needed but not stored. But history is evaluated during the execution of the operation on composed parts, that call the update of the colors.
I made some experiments to identify the changes to be done in the software to solve these problems, and doing this brought me deep into the code and the basic mechanisms on place.
I am convinced there must be a rework of the core to:
-	Avoid immediate calculation after propagation of the changes,
-	Avoid multiple calculation,
-	Avoid unnecessary calculation.
These change would lead to drastic improvement on performances, especially when animating a model. I really think it would really improve user experience and stability of the software, and anyway, without this change, the job around colors cannot be completed.
I have some precise thoughts about what should be done and how it should be done, but as this concerns the core the decisions should be made in collaboration with the designers of the software to be certain to address all aspects.
I noticed also that it is very long to convince to take PR with quite limited and circumscribed changes in the software that correct errors of the actual software. 
I am therefore certain a change that would touch the core would therefore almost impossible to have accepted, what is normal if this change is pushed with no preparation. It is also obvious that, as the software is quite vast, all the consequences must have been discussed to envisage all consequences and side effects. 
So that is why, unless we can setup a discussion on what could be the changes with the designers and have some sprints to experiment these changes, it is not worth for me spending time on that. So that is why I think this PR as it is should be closed.
To continue on that subject a discussion to decide the changes in the design needed should be set, and I am ready and would be glad to participate to such a discussion around the events management topic and behavior of the software. I would also be available to implement whatever could come out these discussions.
Regards
Pascal Garcia

**Pull request closed without merging**